infrastructures/docker/infrastructure/src/main/java/org/eclipse/che/workspace/infrastructure/docker/local/projects/LocalProjectsMigrator.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright (c) 2012-2017 Red Hat, Inc.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.che.workspace.infrastructure.docker.local.projects;
+
+import static org.eclipse.che.api.core.Pages.iterate;
+
+import com.google.inject.Inject;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.nio.file.StandardCopyOption;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Map.Entry;
+import org.eclipse.che.api.core.ServerException;
+import org.eclipse.che.api.workspace.server.model.impl.WorkspaceImpl;
+import org.eclipse.che.api.workspace.server.spi.WorkspaceDao;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Performs migration of projects that are stored for workspace in directories, named respectively
+ * by their names, to new
+ *
+ * @author Mykhailo Kuznietsov
+ */
+public class LocalProjectsMigrator {
+
+  private final WorkspaceDao workspaceDao;
+
+  @Inject
+  public LocalProjectsMigrator(WorkspaceDao workspaceDao) {
+    this.workspaceDao = workspaceDao;
+  }
+
+  private static final Logger LOG = LoggerFactory.getLogger(LocalProjectsMigrator.class);
+
+  public void performMigration(String workspaceProjectsRootFolder) {
+    Map<String, String> workspaceName2id = getId2NameWorkspaceMapping();
+    for (Entry<String, String> entry : workspaceName2id.entrySet()) {
+      Path workspaceStoredByNameLocation =
+          Paths.get(workspaceProjectsRootFolder).resolve(entry.getValue());
+      if (!Files.exists(workspaceStoredByNameLocation)) {
+        // migration is not needed for this workspace
+        continue;
+      }
+      LOG.info(
+          "Performing migration of workspace with id '{}' and name '{}'",
+          entry.getKey(),
+          entry.getValue());
+      Path workspaceStoredByIdLocation =
+          Paths.get(workspaceProjectsRootFolder).resolve(entry.getKey());
+      try {
+        Files.move(
+            workspaceStoredByNameLocation,
+            workspaceStoredByIdLocation,
+            StandardCopyOption.ATOMIC_MOVE);
+        LOG.info(
+            "Successfully migrated projects of workspace with id '{}' and name '{}'",
+            entry.getKey(),
+            entry.getValue());
+      } catch (IOException e) {
+        LOG.error(
+            "Failed to migrate projects of workspace with id '{}' and name '{}'",
+            entry.getKey(),
+            entry.getValue());
+      }
+    }
+  }
+
+  private Map<String, String> getId2NameWorkspaceMapping() {
+    try {
+      Map<String, String> result = new HashMap<>();
+
+      for (WorkspaceImpl workspace :
+          iterate(
```

Comment:
Iterate will eager fetch all objects from a database. I'd be nice to check how many memory it will consume when a number of workspaces is like 50 000 or the same as we have on `codenvy.io`.
If quite a lot then we should fetch workspace ids and names only. Something like that we've already done for another migration.

Reply:
I think that's unnecessary, I don't think there will ever happen a migration of significant size of workspaces, ever.